### PR TITLE
Review Downloader & Checkstyle

### PR DESCRIPTION
**Pull request by thudoan1706**

### CODE REVIEW

File: src/main/java/org/hydev/mcpm/client/Downloader.java [hunk 1]
```diff
@@ -38,10 +84,25 @@ public void downloadFile(String url, File to)
...
+        if (files.size() > 0) {
+            for (int i=0; i<files.size(); i++) {
+                executor.submit(new Processor(i, urls, files));
+            }
+            executor.shutdown();
+            // All files submitted for downloading
+
+
+            try{
+                executor.awaitTermination(20, TimeUnit.SECONDS);
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/Downloader.java]: Does this mean we'll count files that take longer than 20 seconds to download as a failure?

File: src/main/java/org/hydev/mcpm/client/Downloader.java [hunk 2]
```diff
@@ -67,4 +128,53 @@ public Downloader threads(int threads)
         this.threads = threads;
         return this;
     }
-}
+
+    class Processor implements Runnable {
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/Downloader.java:124]: This is cool! Might be worth splitting this out into its own `Processor.java` file if possible, although I can see this is difficult with the `downloadFile` invocation. Your call.

**Review by 1whatleytay — APPROVED** [review]
LGTM!

File: src/main/java/org/hydev/mcpm/client/Downloader.java [hunk 3]
```diff
@@ -25,8 +45,34 @@
      */
     public void downloadFile(String url, File to)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        try (FileOutputStream fileos = new FileOutputStream(to)) {
+            URL link = new URL(url);
+            HttpURLConnection http = (HttpURLConnection)link.openConnection();
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/Downloader.java]: Also still some check style stuff to do if you're interested in the green pr checkmark. 
```suggestion
            HttpURLConnection http = (HttpURLConnection) link.openConnection();
```

```
19
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:42: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[20](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:21)
> Task :checkstyleMain
[21](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:22)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:69: ';' is not followed by whitespace. [WhitespaceAfter]
[22](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:23)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:70: WhitespaceAround: '}' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[23](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:24)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:70: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[24](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:25)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:70: '}' at column 70 should be alone on a line. [RightCurlyAlone]
[25](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:26)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Launcher.java:10:71: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[26](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:27)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:50:56: 'typecast' is not followed by whitespace. [WhitespaceAfter]
[27](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:28)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:51:34: 'typecast' is not followed by whitespace. [WhitespaceAfter]
[28](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:29)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:91:23: WhitespaceAround: '=' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[29](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:30)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:91:23: WhitespaceAround: '=' is not preceded with whitespace. [WhitespaceAround]
[30](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:31)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:91:28: WhitespaceAround: '<' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[31](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:32)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:91:28: WhitespaceAround: '<' is not preceded with whitespace. [WhitespaceAround]
[32](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:33)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:98:13: 'try' is not followed by whitespace. [WhitespaceAfter]
[33](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:34)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:98:13: WhitespaceAround: 'try' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[34](https://github.com/CSC207-2022F-UofT/mcpm/actions/runs/3501674813/jobs/5865523362#step:6:35)
[ant:checkstyle] [WARN] /home/runner/work/mcpm/mcpm/src/main/java/org/hydev/mcpm/client/Downloader.java:98:16: WhitespaceAround: '{' is not preceded with whitespace. [WhitespaceAround]
```

**Review by MstrPikachu — APPROVED** [review]
a bit of checkstyle but it's good

File: src/main/java/org/hydev/mcpm/client/Downloader.java (hunk 1 above)
Comment (thudoan1706) [reply src/main/java/org/hydev/mcpm/client/Downloader.java]: Oh, it meant we wanted to wait for the downloading file complete in the thread for the 20s before executing the following body method can execute.

Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/client/Downloader.java]: What if the thread takes longer than 20 seconds?

File: src/main/java/org/hydev/mcpm/client/Downloader.java (hunk 2 above)
Comment (thudoan1706) [reply src/main/java/org/hydev/mcpm/client/Downloader.java:124]: How though for "downloadFile invocation"?

File: src/main/java/org/hydev/mcpm/client/Downloader.java (hunk 1 above)
Comment (thudoan1706) [reply src/main/java/org/hydev/mcpm/client/Downloader.java]: Then it's doom. But I saw there is a suggestion to use the following to make sure threads finish.

while (!executor.isTerminated()) {}

**Review by thudoan1706 — COMMENTED** [review]
fix check_style

File: src/main/java/org/hydev/mcpm/client/Downloader.java (hunk 2 above)
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/client/Downloader.java:124]: This can be probably a lambda actually:
```
executor.submit(() -> {
  String url = files.get(id);
  File to = urls.get(url);
  downloadFile(url, to);
});
```

File: src/main/java/org/hydev/mcpm/client/Launcher.java [hunk 4]
```diff
@@ -7,9 +7,4 @@
  * @since 2022-09-27
  */
 public class Launcher
-{
-    public static void main(String[] args)
-    {
-        System.out.println("Meow");
-    }
-}
+{ public static void main(String[] args) {System.out.println("Meow");}}
```
Comment (hykilpikonna) [inline src/main/java/org/hydev/mcpm/client/Launcher.java:10]: why is this file changed lol

**Review by hykilpikonna — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/Launcher.java (hunk 4 above)
Comment (thudoan1706) [reply src/main/java/org/hydev/mcpm/client/Launcher.java:10]: I just wanna borrow it for testing lmao and happen to change it hehe

File: src/main/java/org/hydev/mcpm/client/Downloader.java (hunk 2 above)
Comment (thudoan1706) [reply src/main/java/org/hydev/mcpm/client/Downloader.java:124]: oooh nice. Let's me see.

**Comment by hykilpikonna** [comment]
https://user-images.githubusercontent.com/22280294/202871826-8ec975c9-a3ef-4d94-828d-1f21bae9ceda.mp4

Works!

File: src/main/java/org/hydev/mcpm/client/Launcher.java (hunk 4 above)
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/Launcher.java:10]: I'll fix this on main

File: src/main/java/org/hydev/mcpm/client/Downloader.java [hunk 5]
```diff
@@ -25,8 +42,34 @@
...
+                    .descLen(30);
+            bar.appendBar(row);
+            allRows.add(row);
+            while ((read = in.read(buffer, 0, 1024)) >= 0) {
+                bout.write(buffer, 0, read);
+                allRows.forEach(a -> a.increase(1024));
+            }
+            bout.close();
+            in.close();
+            System.out.println("Download completed");
```
Comment (aanushkasharma) [inline src/main/java/org/hydev/mcpm/client/Downloader.java:68]: There appears to be an issue here where, when downloading a single file, it prints multiple progress bars each time a new percentage is reached. In attempt to fix this, I went into ProgressBar.java and went to the forceUpdate method, and changed `out.println(bar.toString(theme, cols)); ` to `out.print(bar.toString(theme, cols)); `. This way it update the same progress bar each time. However, I wasn't sure if this would cause issues for the other classes/methods using ProgressBar.

Comment (aanushkasharma) [reply src/main/java/org/hydev/mcpm/client/Downloader.java:68]: If this works, you might also want to change `System.out.println("Download completed");` to `System.out.println("\nDownload completed");` so that Download Completed doesn't appear right beside the progress bar.

File: src/main/java/org/hydev/mcpm/client/Downloader.java [hunk 6]
```diff
@@ -67,4 +120,53 @@ public Downloader threads(int threads)
...
+        // Remember to chang link to test
+        String link = "https://sd.blackball.lv/library/Introduction_to_Algorithms_Third_Edition_(2009).pdf";
+        File out = new File("./Introduction_to_Algorithms_Third_Edition.pdf");
+        String link1 = "https://www.iusb.edu/students/academic-success-programs/academic-centers-for-excellence/docs/Basic%20Math%20Review%20Card.pdf";
+        File out1 = new File("./Math.pdf");
+        Downloader downloader = new Downloader();
+        Map<String, File> urls = new HashMap<>();
+        urls.put(link, out);
+        urls.put(link1, out1);
+        downloader.downloadFiles(urls, true, 2);
```
Comment (aanushkasharma) [inline src/main/java/org/hydev/mcpm/client/Downloader.java:170]: I tested this with three files and two threads, just to ensure that it doesn't slow down significantly when there are more than 2 files being downloaded, and it seems to work perfectly :)
